tools/buildsys/src/deps.rs

```diff
@@ -0,0 +1,196 @@
+use toml::Table;
+
+// TODO - move this to the Manifest struct
+/// List all packages that are package dependencies. That is, follow all dependencies in the cargo
+/// dependency graph that lead to more packages, and do not follow those that involve kits. This
+/// gives a list of all the packages that are required when we are build a package, or all of the
+/// packages that should be included when building a kit.
+pub fn package_dependencies(manifest_path: impl AsRef<Path>) -> Result<Vec<String>> {
+    let manifest_path = manifest_path.as_ref();
+    let (name, manifest_type) = get_name_and_type(&manifest_path).unwrap();
+    // TODO - deserialize instead of using a MetadataCommand.
+    let graph = MetadataCommand::new()
+        .manifest_path(&manifest_path)
+        .build_graph()
+        .unwrap();
+    let id = find_id(&name, &graph, manifest_type).unwrap();
+    let ids = [&id];
+    let query = graph.query_forward(ids.into_iter()).unwrap();
+    let something = query.resolve_with_fn(|query, link| {
+        let to = link.to();
+        is_manifest_type(&to, ManifestType::Package)
+    });
+    let mut packages: Vec<String> = something
+        .packages(DependencyDirection::Forward)
+        .filter_map(|package_metadata| {
+            // The first dependency is always included, and if it happens to be something other than
+            // a package, we want to filter it out.
+            if is_manifest_type(&package_metadata, ManifestType::Package) {
+                // Return the name instead of the metadata or package ID.
+                Some(package_metadata.name().to_string())
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // Sort so that this function has consistent, dependable output regardless of graph internals.
+    packages.sort();
+    Ok(packages)
+}
+
+// TODO - move this to the Manifest struct
+/// List all kits needed to build a package or variant.
+pub fn kit_dependencies(manifest_path: impl AsRef<Path>) -> Result<Vec<String>> {
+    let manifest_path = manifest_path.as_ref();
+    // TODO - deserialize instead of using a MetadataCommand.
+    let mut kits: Vec<String> = MetadataCommand::new()
+        .manifest_path(&manifest_path)
+        .build_graph()
+        .unwrap()
+        .packages()
+        .filter_map(|package_metadata| {
+            if is_manifest_type(&package_metadata, ManifestType::Kit) {
+                Some(package_metadata.name().to_string())
+            } else {
+                None
+            }
+        })
+        .collect();
+    kits.sort();
+    Ok(kits)
+}
+
+fn is_manifest_type(package_metadata: &PackageMetadata, manifest_type: ManifestType) -> bool {
+    let metadata_table = package_metadata.metadata_table();
+    match manifest_type {
+        ManifestType::Package => metadata_table.get("build-package").is_some(),
+        ManifestType::Kit => metadata_table.get("build-kit").is_some(),
+    }
+}
```

Comment:
Not sure where this fix would go, but we have the problem today that the manifest is very loose and will allow both `build-package` and `build-variant` tables to exist in the same file without throwing an error. So we might end up confused here if there's a malformed Cargo.toml.

Reply:
> Not sure where this fix would go, but we have the problem today that the manifest is very loose and will allow both build-package and build-variant tables to exist in the same file without throwing an error. So we might end up confused here if there's a malformed Cargo.toml.

I'm thinking I want to defer this tightening just due to the amount of code I'm about to bomb.